### ISSUE TITLE
How to use with ABC / ABCMeta

### ISSUE THREAD

**Issue by jcugat** [samuelcolvin/pydantic#120]
I was trying to define a pydantic class with abstract methods, which was harder than I thought due to metaclass conflicts. I finally managed to do it like this:

```python
from abc import ABCMeta, abstractmethod
from pydantic import BaseModel
from pydantic.main import MetaModel

class PydanticWithABCMeta(ABCMeta, MetaModel):
    pass

class MyPydanticClass(BaseModel, metaclass=PydanticWithABCMeta):
    some_field: str

    @abstractmethod
    def my_abstract_method(self):
        pass
```

Not sure it's the best approach, but if it is, would be worth documenting it somewhere?

**Comment by samuelcolvin** [comment]
Humm, I've never worked with abc myself, but I see no reason not to add `ABCMeta` to the normal metaclass and thereby allow `@abstractmethod` to be used with minimum fuss.

Are there any downsides to adding `ABCMeta`? Eg. could it effect performance in any way?

Happy to accept a PR to either add documentation or `ABCMeta`.

**Comment by jcugat** [comment]
I didn't find any meaningful performance hit.

Here's the script I used:
```python
import timeit

setup_bare = '''
from typing import List

from pydantic import BaseModel

class MyPydanticClass(BaseModel):
    id: int
    name: str
    friends: List[int]
'''

setup_abc = '''
from abc import ABCMeta, abstractmethod
from typing import List

from pydantic import BaseModel
from pydantic.main import MetaModel

class PydanticWithABCMeta(ABCMeta, MetaModel):
    pass

class MyPydanticClass(BaseModel, metaclass=PydanticWithABCMeta):
    id: int
    name: str
    friends: List[int]
'''

stmt = 'MyPydanticClass(id=1, name="name", friends=[1,2,3,4])'

print('Pydantic base class:')
print(timeit.Timer(stmt, setup=setup_bare).repeat(5))
print('Pydantic ABCMeta subclass:')
print(timeit.Timer(stmt, setup=setup_abc).repeat(5))
```

And the results in my machine:
```bash
$ python pydantic_benchmark.py
Pydantic base class:
[18.120516726994538, 18.072193423999124, 18.320782110997243, 18.08288959201309, 18.0655533040117]
Pydantic ABCMeta subclass:
[17.97471183000016, 17.94070240900328, 17.806602609998663, 17.75709401500353, 18.035982765999506]
$ python -VV
Python 3.6.3 (default, Dec 12 2017, 13:52:48)
[GCC 4.2.1 Compatible Apple LLVM 8.1.0 (clang-802.0.42)]
$ pip freeze
pydantic==0.6.4
```

**Comment by samuelcolvin** [comment]
Cool. I think we add `ABCMeta` to `MetaModel` and then add documentation on how to use it. Would you like to submit a PR?

By the way: pydantic has a fairly comprehensive set of benchmarks along with comparisons of other similar libraries: https://github.com/samuelcolvin/pydantic/tree/master/benchmarks.

**Comment by jcugat** [comment]
Addressed in #123

**Comment by samuelcolvin** [comment]
thanks for your work on this.
